File: runner.pl
#!/usr/bin/env perl
# -*- mode: cperl; indent-tabs-mode: nil; tab-width: 3; cperl-indent-level: 3; -*-
use strict;
use warnings;
use utf8;

BEGIN {
	$| = 1;
	binmode(STDIN, ':encoding(UTF-8)');
	binmode(STDOUT, ':encoding(UTF-8)');
}
use open qw( :encoding(UTF-8) :std );
use feature 'unicode_strings';
use Digest::SHA qw(sha1_base64);
use File::Basename;
use POSIX;

if ((($ENV{LANGUAGE} || "").($ENV{LANG} || "").($ENV{LC_ALL} || "")) !~ /UTF-?8/i) {
   die "Locale is not UTF-8 - bailing out!\n";
}
if ($ENV{PERL_UNICODE} !~ /S/ || $ENV{PERL_UNICODE} !~ /D/ || $ENV{PERL_UNICODE} !~ /A/) {
   die "Envvar PERL_UNICODE must contain S and D and A!\n";
}

use FindBin qw($Bin);
use lib "$Bin/";
use Helpers;

use Getopt::Long;
Getopt::Long::Configure('no_ignore_case');
my %opts = ();
GetOptions(\%opts,
   'help|h|?',
   'binary|b=s',
   'folder|f=s',
   'corpus|c=s',
   );

if (defined $opts{'binary'}) {
   if (!(-e $opts{'binary'} && -x $opts{'binary'})) {
      print "Error: $opts{'binary'} is not executable!\n";
      exit(1);
   }
   if ($opts{'binary'} !~ m@^[./]@) {
      $opts{'binary'} = './'.$opts{'binary'};
   }
}
else {
   my @bins = glob('./*.pl');
   if (scalar(@bins)) {
      $opts{'binary'} = $bins[0];
   }
}

if (defined $opts{'folder'}) {
   if (!(-e $opts{'folder'} && -d $opts{'folder'} && -w $opts{'folder'})) {
      print "Error: $opts{'folder'} is not a writable folder!\n";
      exit(1);
   }
}
else {
   $opts{'folder'} = 'regtest';
}

if (defined $opts{'binary'} && substr($opts{'folder'}, 0, 1) ne '/') {
   $opts{'folder'} = dirname($opts{'binary'}).'/'.$opts{'folder'};
}

my $corpus = '';
if (defined $opts{'corpus'}) {
   $corpus = $opts{'corpus'};
}
elsif (defined $ARGV[0]) {
   $corpus = $ARGV[0];
}

if (defined $opts{'help'}) {
   my @cns = ();
   my @fs = glob("$opts{'folder'}/input-*.txt");
   foreach my $f (@fs) {
      my ($bn) = ($f =~ m@\Q$opts{'folder'}\E/input-(\S+?).txt@);
      push(@cns, $bn);
   }

   print "analyse.pl\n";
   print "\t--help, -h, -?  Show this help\n";
   print "\t--binary, -b    Path to the binary to get regression test pipe from; defaults to the first found *.pl file in current working dir.\n";
   print "\t--folder, -f    Path to the folder containing the input files; defaults to ./regtest/ relative to the binary's folder.\n";
   print "\t--corpus, -c    Name of the corpus to run; defaults to all found corpora.\n";
   print "\n";
   print "Possible corpus names:\n\t".join("\n\t", @cns)."\n";
   exit(0);
}

$ENV{REGRESSION_TEST} = 1;

my $cmd_run_p = trim(`'$opts{'binary'}' --regtest`);
my $cmd_raw = trim(`'$opts{'binary'}' --regtest --raw`);

if ($cmd_run_p !~ / REGTEST_/ || $cmd_raw !~ / REGTEST_/) {
   print "Error: Binary did not return a usable regression test pipe!\n";
   exit(1);
}

my @fs = glob("$opts{'folder'}/input-*.txt");
foreach my $f (@fs) {
   my ($bn) = ($f =~ m@\Q$opts{'folder'}\E/input-(\S+?).txt@);

   if ($corpus && $bn !~ /^$corpus$/) {
      next;
   }

   print "Handling $bn ...\n";
   print "\tinput $f\n";
   `rm -rf '$opts{'folder'}/output-$bn-'*`;

   my $i = 0;
   my %uniq = ();
   my @sents = ();

   my $data = file_get_contents($f);
   $data =~ s@\x{feff}@@g; # Unicode BOM
   $data =~ s@<STREAMCMD:FLUSH>@@g;
   $data =~ s@[ \t]+\n@\n@g;
   $data =~ s@\r\n@\n@g;
   $data =~ s@\r@\n@g;

   my @ins = ();
   if ($data =~ m@(^|\n)<s[^>]*>(\n|$)@ && $data =~ m@\n"<[^\n]+>"@) {
      print "\tdelimiting by <s> tags\n";
      my @lns = split(/\n<\/s[^>]*>/, $data);
      foreach (@lns) {
         ++$i;
         s@(^|\n)<s[^>]*>@@g;
         s@\n\n+@\n\n@g;
         $_ = trim($_);
         if (!$_) {
            # Skip empty
            next;
         }
         push(@ins, [$i, $_]);
      }
   }
   else {
      print "\tdelimiting by lines\n";
      my @lns = split(/\n+/, $data);
      foreach (@lns) {
         ++$i;
         $_ =~ s/#[^\n]*//g;
         $_ = trim($_);
         $_ =~ s/\s\s+/ /g;
         $_ =~ s/\\n/\n/g;
         $_ =~ s/[ \t]+\n/\n/g;
         if (!$_ || /^</) {
            # Skip empty or commented lines
            next;
         }
         push(@ins, [$i, $_]);
      }
   }

   foreach (@ins) {
      my ($i, $s) = @$_;
      $_ = $s;
      utf8::encode($s); # sha1_base64() can't handle UTF-8 for some reason
      my $hash = sha1_base64($s);
      $hash =~ s/[^a-zA-Z0-9]/x/g;
      if (defined $uniq{$hash}) {
         # Skip duplicate inputs
         next;
      }
      $uniq{$hash} = 1;
      push(@sents, "<s$hash-$i>\n".$_."\n</s$hash-$i>\n\n<STREAMCMD:FLUSH>");
   }

   `rm -rf '$opts{'folder'}/cmd-$bn-'*`;

   @sents = sort(@sents);
   my $len = scalar @sents;
   for (my $i=0 ; $i<4 ; ++$i) {
      file_put_contents("$opts{'folder'}/output-$bn-010.txt.$i", join("\n\n", splice(@sents, 0, ceil($len/4))));

      my $cmd_run = $cmd_run_p;
      while ($cmd_run =~ / REGTEST_(\S+) (\S+)/) {
         my ($type,$opt) = ($1, $2);
         my $rpl = "| tee '$opts{'folder'}/output-$bn-$opt.txt.$i'";
         if ($type eq 'CG') {
            $rpl = "| cg-sort | tee '$opts{'folder'}/output-$bn-$opt-trace.txt.$i' | cg-untrace | cg-sort | tee '$opts{'folder'}/output-$bn-$opt.txt.$i'";
         }
         $cmd_run =~ s/\| REGTEST_\S+ \S+s*/$rpl/;
      }
      file_put_contents("$opts{'folder'}/cmd-$bn-run.$i", $cmd_run);
   }

   file_put_contents("$opts{'folder'}/cmd-$bn-raw", $cmd_raw);

   my $cmd_par = <<XCMD;
cat '$opts{'folder'}/output-$bn-010.txt.0' | bash '$opts{'folder'}/cmd-$bn-run.0' >'$opts{'folder'}/output-$bn.0' 2>'$opts{'folder'}/error-$bn' &
cat '$opts{'folder'}/output-$bn-010.txt.1' | bash '$opts{'folder'}/cmd-$bn-run.1' >'$opts{'folder'}/output-$bn.1' 2>'$opts{'folder'}/error-$bn' &
cat '$opts{'folder'}/output-$bn-010.txt.2' | bash '$opts{'folder'}/cmd-$bn-run.2' >'$opts{'folder'}/output-$bn.2' 2>'$opts{'folder'}/error-$bn' &
cat '$opts{'folder'}/output-$bn-010.txt.3' | bash '$opts{'folder'}/cmd-$bn-run.3' >'$opts{'folder'}/output-$bn.3' 2>'$opts{'folder'}/error-$bn' &

for job in `jobs -p`
do
	wait \$job
done

cat '$opts{'folder'}/output-$bn-010.txt.0' '$opts{'folder'}/output-$bn-010.txt.1' '$opts{'folder'}/output-$bn-010.txt.2' '$opts{'folder'}/output-$bn-010.txt.3' > '$opts{'folder'}/output-$bn-010.txt'

cat '$opts{'folder'}/output-$bn.0' '$opts{'folder'}/output-$bn.1' '$opts{'folder'}/output-$bn.2' '$opts{'folder'}/output-$bn.3'

rm -rf '$opts{'folder'}/output-$bn.'*
rm -rf '$opts{'folder'}/output-$bn-010.txt.'*

XCMD
   file_put_contents("$opts{'folder'}/cmd-$bn-run-parallel", $cmd_par);
   `nice -n20 bash '$opts{'folder'}/cmd-$bn-run-parallel'`;
   print "\n";

   for (my $i=0 ; $i<4 ; ++$i) {
      my $cmd_run = $cmd_run_p;
      while ($cmd_run =~ / REGTEST_(\S+) (\S+)/g) {
         my ($type,$opt) = ($1, $2);
         `cat '$opts{'folder'}/output-$bn-$opt.txt.$i' >> '$opts{'folder'}/output-$bn-$opt.txt'`;
         if ($type eq 'CG') {
            `cat '$opts{'folder'}/output-$bn-$opt-trace.txt.$i' >> '$opts{'folder'}/output-$bn-$opt-trace.txt'`;
         }
         unlink("$opts{'folder'}/output-$bn-$opt.txt.$i");
         unlink("$opts{'folder'}/output-$bn-$opt-trace.txt.$i");
      }
   }
}

File: runner.py
#!/usr/bin/env python3
import argparse
import base64
import glob
import hashlib
import json5
import math
import os
import re
import shutil
import subprocess
import sys
import time
from pathlib import Path

import Helpers

def hash(s):
    return base64.b64encode(hashlib.sha1(s.encode('UTF-8')).digest(), b'xx').decode('UTF-8').rstrip('=')

parser = argparse.ArgumentParser(prog='runner.py', description='Regression test runner')
#parser.add_argument('-l', '--list', action='store_true', help='List usable tests and corpora')
parser.add_argument('-P', '--proc', action='store', help='Number of parallel processes; defaults to max(3,num_cores/4)', default=int(max(3, os.cpu_count()/4)))
parser.add_argument('-f', '--folder', action='store', help='Folder with regtest.json5; defaults to looking in ./, regtest/, or test/', default='')
parser.add_argument('-c', '--corp', action='append', help='Restricts the test to the named corpora; can be given multiple times and/or pass a comma separated list', default=[])
parser.add_argument('-D', '--debug', action='store_true', help='Enable Python stack traces and other debugging', default=False)
parser.add_argument('test', nargs='?', help='Which test to run; defaults to first defined', default='')
args = parser.parse_args()

if not args.debug:
	sys.tracebacklimit = 0

root = Helpers.find_root(args.folder)

procs = int(args.proc)

timeout = Helpers.timeout()
timeout_sec = 1800 # Half an hour

config = Helpers.load_config(root)

# Determine which test to run
tkey, test = Helpers.resolve_test(config, args.test)

# Determine which corpora to run
corps = Helpers.resolve_corps(root, config, test, args.corp)

if os.path.exists(f'{root}/output/{tkey}/_tmp/lock'):
	pid = Path(f'{root}/output/{tkey}/_tmp/lock').read_text(encoding='UTF-8')
	ps = subprocess.run(['ps', '-p', pid], capture_output=True, encoding='UTF-8').stdout
	if re.search(rf'\n{pid}.*(python|runner)', ps):
		raise RuntimeError(f'{root}/output/{tkey}/_tmp/lock exists from PID {pid} - potential double-run! Remove _tmp folder or lock file if remnant from a crash.')
	else:
		print(f'Removing stale lock from PID {pid}')

shutil.rmtree(f'{root}/output/{tkey}/_tmp', ignore_errors=True)
os.makedirs(f'{root}/output/{tkey}/_tmp/', exist_ok=True)
Path(f'{root}/output/{tkey}/_tmp/lock').write_text(str(os.getpid()))

steps, _ = Helpers.resolve_steps(config, test)

pipe = ''
for p,s in steps.items():
	# Every step gets its own timeout watcher, to avoid runaway background processes
	pipe += f' | {timeout} {timeout_sec} {s["cmd"]}'
	if 'trace' in s:
		pipe += f' {s["trace"]} 2>{root}/output/{tkey}/_tmp/step-{p}.NNN.err | tee {root}/output/{tkey}/_tmp/step-{p}.trace.NNN'
	elif s['type'] == 'cg':
		pipe += f' --trace 2>{root}/output/{tkey}/_tmp/step-{p}.NNN.err | cg-sort | tee {root}/output/{tkey}/_tmp/step-{p}.trace.NNN | cg-untrace | cg-sort'
	else:
		pipe += f' 2>{root}/output/{tkey}/_tmp/step-{p}.NNN.err'
	pipe += f' | tee {root}/output/{tkey}/_tmp/step-{p}.NNN'
pipe = pipe.replace(' --trace --trace ', ' --trace ')
pipe = pipe.replace(' cg3-autobin.pl ', ' vislcg3 ')
pipe = pipe[3:]

inputs = {}
uniq_inputs = {}
for c,f in corps.items():
	inputs[c] = {}
	with open(f, 'r', encoding='UTF-8') as fd:
		ln = 0
		while l := fd.readline():
			ln += 1
			start = ln
			l = l.rstrip()

			# Read in <s>...</s> segments as a literal block
			if l.startswith('<s ') or l.startswith('<s>'):
				l += '\n'
				while e := fd.readline():
					ln += 1
					l += e
					if e.startswith('</s>'):
						break
				l = re.sub(r'[ \t]+\n', '\n', l)
				l = re.sub(r'\n\n\n+', '\n\n', l)
				l = l.strip()
			else:
				l = re.sub(r'#.*', '', l) # Strip comments
				l = re.sub(r'\s\s+', ' ', l)
				l = re.sub(r'\\n', '\n', l) # Convert literal \n to newline
				l = re.sub(r'[ \t]+\n', '\n', l)
				l = l.strip()
				# Other tag-looking lines are skipped (backwards compat)
				if l.startswith('<'):
					continue

			if not l or l == '':
				continue

			h = hash(l)
			if h not in inputs[c]:
				inputs[c][h] = start
			if h not in uniq_inputs:
				uniq_inputs[h] = l

# Just in case there are fewer inputs than CPU cores
procs = min(procs, len(uniq_inputs))

for c,hs in inputs.items():
	with open(f'{root}/output/{tkey}/corp-{c}.ids', 'w', encoding='UTF-8') as fd:
		for k,v in hs.items():
			fd.write(f'{k}\t{v}\n')

fds = []
for i in range(procs):
	fds.append(open(f'{root}/output/{tkey}/_tmp/input.{i}', 'w', encoding='UTF-8'))

chunk = int(len(uniq_inputs)/procs)+1
for i,k in enumerate(sorted(uniq_inputs.keys())):
	t = uniq_inputs[k]
	if t.startswith('<s'):
		t = re.sub(r'^<s', f'<s id="{k}"', t)
	else:
		t = f'<s id="{k}">\n{t}\n</s>'
	fds[min(int(math.floor(i/chunk)), procs-1)].write(f'{t}\n\n<STREAMCMD:FLUSH>\n\n')

bash = ''
for e in test['env']:
	bash += f'export "{e}"\n'

outs = []
for i in range(procs):
	fds[i].close()
	np = re.sub(r'\.NNN', f'.{i}', pipe)
	Path(f'{root}/output/{tkey}/_tmp/sh.{i}').write_text(np)
	# Race prevention: Create and open the output files for reading, so the script can append to existing files
	Path(f'{root}/output/{tkey}/_tmp/out.{i}').touch()
	outs.append(open(f'{root}/output/{tkey}/_tmp/out.{i}', 'r', encoding='UTF-8'))
	bash += f'cat {root}/output/{tkey}/_tmp/input.{i} | bash {root}/output/{tkey}/_tmp/sh.{i} >>{root}/output/{tkey}/_tmp/out.{i} 2>{root}/output/{tkey}/_tmp/err.{i} &\n'

bash += f'''
for job in `jobs -p`
do
	#echo "Waiting for $job"
	wait $job
done

echo "Done" > {root}/output/{tkey}/_tmp/done
'''

Path(f'{root}/output/{tkey}/_tmp/sh').write_text(bash)
seen = set()
proc = subprocess.Popen([timeout, str(timeout_sec), 'nice', '-n20', 'bash', f'{root}/output/{tkey}/_tmp/sh'])

print('Running: %s -P %s -f %s -c %s %s' % (os.path.relpath(__file__), str(procs), root, ','.join(sorted(corps.keys())), tkey))
print('Progress: 0%', end='\r', flush=True)
while not proc.poll():
	time.sleep(1)
	did = False
	for i in range(procs):
		while l := outs[i].readline():
			if m := re.search(r'^<s id="([^"]+)"', l):
				seen.add(m[1])
				did = True
	if did:
		print('Progress: {}%'.format(int(len(seen) / len(uniq_inputs) * 100)), end='\r', flush=True)

	if os.path.exists(f'{root}/output/{tkey}/_tmp/done'):
		break

print('Progress: 100%', end='\r', flush=True)

for i in range(procs):
	outs[i].close()

if len(seen) != len(uniq_inputs):
	# Check again, because sometimes the above loop skips an ID
	seen = set()
	for i in range(procs):
		seen |= set(re.findall(r'(?:^|\n)<s id="([^"]+)"', Path(f'{root}/output/{tkey}/_tmp/out.{i}').read_text()))
	if len(seen) != len(uniq_inputs):
		missing = set(uniq_inputs.keys()) - set(seen)
		print('Warning: Missing outputs - got {0} of {1}! Example missing ID: {2}'.format(len(seen), len(uniq_inputs), list(missing)[0]))

for c in inputs.keys():
	shutil.rmtree(f'{root}/output/{tkey}/{c}', ignore_errors=True)
	os.makedirs(f'{root}/output/{tkey}/{c}/', exist_ok=True)

def split_to_corps(s, f):
	global procs, inputs, root, test

	cfs = {}
	for c in inputs.keys():
		cfs[c] = open(f'{root}/output/{tkey}/{c}/output-{c}-{s}.txt', 'w')

	for i in range(procs):
		with open(f'{f}.{i}', 'r', encoding='UTF-8') as fd:
			while l := fd.readline():
				if l.startswith('<s '):
					while e := fd.readline():
						l += e
						if e.startswith('</s>'):
							break
					l = re.sub(r'[ \t]+\n', '\n', l)
					l = re.sub(r'\n\n\n+', '\n\n', l)
					m = re.search(r'^<s id="([^"]+)"', l)
					for c in inputs.keys():
						if m[1] in inputs[c]:
							cfs[c].write(l + '\n')

	for c in inputs.keys():
		cfs[c].close()

split_to_corps('010', f'{root}/output/{tkey}/_tmp/input')

for p,s in steps.items():
	t = 'auto'
	if 'type' in s:
		t = s['type']
	if 'trace' in s or t == 'cg':
		split_to_corps(f'{p}-trace', f'{root}/output/{tkey}/_tmp/step-{p}.trace')
	split_to_corps(p, f'{root}/output/{tkey}/_tmp/step-{p}')

os.remove(f'{root}/output/{tkey}/_tmp/lock')
print('Done           ')
